hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java

```diff
@@ -2616,10 +2617,17 @@ public Integer getWritesFileIdEncoding() {
     return props.getInteger(WRITES_FILEID_ENCODING, HoodieMetadataPayload.RECORD_INDEX_FIELD_FILEID_ENCODING_UUID);
   }
 
-  public boolean isNonBlockingConcurrencyControl() {
-    return getTableType().equals(HoodieTableType.MERGE_ON_READ)
-        && getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()
-        && isSimpleBucketIndex();
+  public boolean needResolveWriteConflict(WriteOperationType operationType) {
+    // Skip to resolve conflict for non bulk_insert operation if using non-blocking concurrency control
+    if (getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()) {
+      if (getTableType().equals(HoodieTableType.MERGE_ON_READ) && isSimpleBucketIndex()) {
+        return operationType == WriteOperationType.UNKNOWN || operationType == WriteOperationType.BULK_INSERT;
+      } else {
+        return true;
```

Comment:
So for `UPSERT` we also needs to resolve conflicts?

Reply:
Can we detect the conflict in such scenario?

```java
t1.bulk_insert.request -----------------------------------------------------------> t1_t5.bulk_insert.commit
                                          t2.delta_commit.request ---------> t2_t3.delta_commit
```

and

```java
t1.bulk_insert.request --------------------------------> t1_t4.bulk_insert.commit
                                          t2.delta_commit.request ---------> t2_t5.delta_commit
```

Reply:
> So for `UPSERT` we also needs to resolve conflicts?
No, `UPSERT` does not need resolve conflict.

> Can we detect the conflict in such scenario?
The first case: t1.commit would fail to commit caused by confliction
The first case: both t1 and t2 could commit successful.

Reply:
I'm confused, for UPSERT it returns true, and that means we need to resolve conflicts right?

Reply:
`UPSERT` would return `false`

Reply:
Can we simplify the logic as:

```java
return is_nb_cc ? operation == BULK_INSERT : true;
```